### PR TITLE
[RLlib] Add example: Pre-train an `RLModule` single-agent, then bring checkpoint into multi-agent setup and continue training.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

So far, we have no example that shows users how to pre-train certain policies and load the checkpoints.

This PR shows users how to pre-train a module in single-agent mode and load its checkpoint in another training run into a MARL setup.

## Related issue number

Related to #44263 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/rl_modules/pretraining_rlm.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    default_timesteps=50000,
+    default_reward=500.0,
+    default_iters=20,
+)
+# Instead use mroe for the pre-training run.
+parser.add_argument(
+    "--stop-iters-pretraining",
+    type=int,
+    default=200,
+    help="The number of iterations to train.",
```
Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:26]:
```suggestion
    help="The number of iterations to pre-train.",
```

File: rllib/examples/rl_modules/pretraining_rlm.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    args = parser.parse_args()
+    # Ensure that the user has set the number of agents.
+    if args.num_agents == 0:
+        raise ValueError(
+            "This pre-training example script requires at least 1 agent. "
+            "Try setting the command line argument `--num-agents` to the "
+            "number of agents you want to use."
+        )
+
+    # Store the user's stopping criteria for the löater training run.
```
Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:48]:
```suggestion
    # Store the user's stopping criteria for the later training run.
```

File: rllib/examples/rl_modules/pretraining_rlm.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    "--stop-iters-pretraining",
+    type=int,
+    default=200,
+    help="The number of iterations to train.",
+)
+parser.add_argument(
+    "--stop-timesteps-pretraining",
+    type=int,
+    default=5000000,
+    help="The number of (environment sampling) timesteps to train.",
```
Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:32]:
```suggestion
    help="The number of (environment sampling) timesteps to pre-train.",
```

File: rllib/examples/rl_modules/pretraining_rlm.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    setattr(args, "stop_iters", args.stop_iters_pretraining)
+    setattr(args, "stop_timesteps", args.stop_timesteps_pretraining)
+    setattr(args, "checkpoint_at_end", True)
+
+    # Define out pre-training single-agent algorithm. We will use the same module
+    # configuration for the pre-training and the training.
+    config = (
+        PPOConfig()
+        # Enable the new API stack (RLModule and Learner APIs).
+        .experimental(_enable_new_api_stack=True)
```
Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:62]: This is done automatically by the `run_rllib_example_script_experiment` util.

Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:62]:
```suggestion
```

File: rllib/examples/rl_modules/pretraining_rlm.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        action_space=env.action_space,
+        model_config_dict={"fcnet_hiddens": [64]},
+        catalog_class=PPOCatalog,
+        # Note, we load here the module directly from the checkpoint.
+        load_state_path=module_chkpt_path,
+    )
+    marl_module_spec = MultiAgentRLModuleSpec(module_specs=module_specs)
+
+    # Register our environment with tune if we use multiple agents.
+    if args.num_agents > 0:
```
Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:101]: Is this if-block needed? We assert that this command line arg is >0 above.

File: rllib/examples/rl_modules/pretraining_rlm.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        register_env(
+            "multi-agent-carpole-env",
+            lambda _: MultiAgentCartPole(config={"num_agents": args.num_agents}),
+        )
+
+    # Configure the main (multi-agent) training run.
+    config = (
+        PPOConfig()
+        # Enable the new API stack (RLModule and Learner APIs).
+        .experimental(_enable_new_api_stack=True)
```
Comment (sven1977) [inline rllib/examples/rl_modules/pretraining_rlm.py:111]:
```suggestion
```

**Review by sven1977 — APPROVED** [review]
Super nice example and PR! Thanks @simonsays1980 !
Just a few nits and waiting for:

1) We must add this great example to the BUILD!
2) Can we rename the script into a more descriptive name? Like `pretraining_single_agent_training_multi_agent` <- something like this that more describes the exact sequence of things we do here.

File: rllib/examples/rl_modules/pretraining_rlm.py (hunk 5 above)
Comment (simonsays1980) [reply rllib/examples/rl_modules/pretraining_rlm.py:99]: Yeah I guess we can remove this here. Good catch @sven1977 !

Comment (simonsays1980) [reply rllib/examples/rl_modules/pretraining_rlm.py:99]: Great catch! I removed this in the follow-up commit.

**Comment by sven1977** [comment]
Ok, cool! Can we also add this example script to BUILD?

File: rllib/BUILD [hunk 7]
```diff
@@ -2873,7 +2873,14 @@ py_test(
     size = "small",
     srcs = ["examples/rl_modules/classes/mobilenet_rlm.py"],
 )
-
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:2876]: Awesome!

File: rllib/utils/test_utils.py [hunk 8]
```diff
@@ -76,7 +76,7 @@ def add_rllib_example_script_args(
     """Adds RLlib-typical (and common) examples scripts command line args to a parser.
 
     TODO (sven): This function should be used by most of our examples scripts, which
-     already mostly have this logic in them (but written out).
+    already mostly have this logic in them (but written out).
```
Comment (sven1977) [inline rllib/utils/test_utils.py:79]:
```suggestion
     already mostly have this logic in them (but written out).
```

**Pull request merged**